blockprod/src/detail/mod.rs

```diff
@@ -73,6 +83,57 @@ impl BlockProductionInterface for BlockProduction {
     fn is_connected(&self) -> bool {
         !self.builder_tx.is_closed()
     }
+
+    async fn generate_block(
+        &self,
+        reward_destination: Destination,
+        signed_transactions: Vec<SignedTransaction>,
+    ) -> Result<Id<Block>, BlockProductionError> {
+        let (best_block_id, best_block_height, accumulator) = self
+            .chainstate_handle
+            .call(|this| {
+                // Stay with DefaultTxaccumulator as we are being
+                // explicit with the transactions to build blocks with
+                let mut accumulator =
+                    DefaultTxAccumulator::new(this.get_chain_config().max_block_size_from_txs());
+
+                for signed_transaction in signed_transactions {
+                    if accumulator.add_tx(signed_transaction, Fee::new(Amount::ZERO)).is_err() {
+                        return None;
+                    };
+                }
+
+                if let Ok(block_id) = this.get_best_block_id() {
+                    if let Ok(Some(block_height)) = this.get_block_height_in_main_chain(&block_id) {
+                        return Some((block_id, block_height, accumulator));
+                    }
+                }
+
+                None
+            })
+            .await?
+            .ok_or(BlockProductionError::SubsystemCallError(
+                CallError::SubmissionFailed,
+            ))?;
+
+        let (_tx, dummy_rx) = crossbeam_channel::unbounded();
+
+        let block_maker = BlockMaker::new(
+            self.chain_config.clone(),
+            self.chainstate_handle.clone(),
+            self.mempool_handle.clone(),
+            self.time_getter.clone(),
+            reward_destination,
+            best_block_id,
+            best_block_height,
+            dummy_rx,
+        );
+
+        // TODO: Stop any current block makers before making this new
+        // block, and then restart if enabled when done. Otherwise we
+        // will be doing work for no benefit.
+        Ok(block_maker.make_block(best_block_id, &accumulator)?.get_id())
```

Comment:
I think the better way to handle this is to change the argument of make_block to take an enum, which can either be a Vec of transactions or an accumulator. The way it's done now is kind of twisted for no good reason; especially that this function is made for testing, so we need behavior that is decoupled from the behavior of the accumulator.

Reply:
Now that `generate_block()` is async, Accumulators can't be passed into the function because they are not `Sync`. Also, we're now passing in by value.